helix-core/src/main/java/org/apache/helix/controller/rebalancer/AbstractRebalancer.java

```diff
@@ -385,10 +389,19 @@ public static int getStateCount(String state, StateModelDefinition stateModelDef
           // If the desired state is the top state, but the instance cannot be transited to the
           // top state in one hop, try to keep the top state on current host or a host with a closer
           // state.
-          for (String currentStatePrioritizedInstance : sortedPreferenceList) {
-            if (!assigned.contains(currentStatePrioritizedInstance) && liveAndEnabled
-                .contains(currentStatePrioritizedInstance)) {
+          while (currentStatePrioritizedInstanceIter.hasNext()) {
+            // Note that it is safe to check the prioritized instance items only once here.
+            // Since the only possible condition when we don't use an instance in this list is that
+            // it has been assigned with some state. And this is not revertable in this method. So
+            // checking it one more time later will only waste time.
+            String currentStatePrioritizedInstance = currentStatePrioritizedInstanceIter.next();
+            if (!assigned.contains(currentStatePrioritizedInstance)) {
```

Comment:
I moved this logic above. Inactive nodes have been removed from the currentStatePrioritizedInstanceList. This shortens the instance list and saves some computing.

Note that we may don't want to do it for the preferencelist since it is not deepcopy.